### PR TITLE
Optimization routines for SHA3

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/SolidityOptimizer.cpp [hunk 1]
```diff
@@ -568,6 +581,155 @@ BOOST_AUTO_TEST_CASE(cse_jumpi_jump)
...
+		Instruction::MSTORE, // m[128] = DUP1
+		u256(0x20),
+		Instruction::DUP1,
+		Instruction::DUP3,
+		Instruction::SHA3, // sha3(m[128..(128+32)])
+		u256(12),
+		Instruction::DUP5,
+		Instruction::DUP2,
+		Instruction::MSTORE, // m[12] = DUP1
+				Instruction::DUP12,
```
Comment (gavofyork) [inline test/SolidityOptimizer.cpp]: weird indentation going on here...

File: libevmcore/CommonSubexpressionEliminator.cpp [hunk 2]
```diff
@@ -194,26 +196,27 @@ ExpressionClasses::Id CommonSubexpressionEliminator::initialStackElement(int _st
...
+void CommonSubexpressionEliminator::storeInStorage(Id _slot, Id _value)
 {
 	if (m_storageContent.count(_slot) && m_storageContent[_slot] == _value)
 		// do not execute the storage if we know that the value is already there
 		return;
 	m_sequenceNumber++;
 	decltype(m_storageContent) storageContents;
-	// copy over values at points where we know that they are different from _slot
+	// copy over values at points where we know that they are different from _slot or that we know
+	// that the value is equal
```
Comment (LefterisJP) [inline libevmcore/CommonSubexpressionEliminator.cpp]: or where we know that the values are equal?

File: libevmcore/ExpressionClasses.h [hunk 3]
```diff
@@ -78,6 +78,15 @@ class ExpressionClasses
 	bool knownToBeDifferent(Id _a, Id _b);
 	/// Similar to @a knownToBeDifferent but require that abs(_a - b) >= 32.
 	bool knownToBeDifferentBy32(Id _a, Id _b);
+	/// @returns true if the value of the given class is known to be zero.
+	/// @note that this is not the negariot of knownNonZero
```
Comment (LefterisJP) [inline libevmcore/ExpressionClasses.h]: negariot -> negation ?

File: libevmcore/ExpressionClasses.h [hunk 4]
```diff
@@ -78,6 +78,15 @@ class ExpressionClasses
...
+	bool knownZero(Id _c);
+	/// @returns true if the value of the given class is known to be nonzero.
+	/// @note that this is not the negariot of knownZero
```
Comment (LefterisJP) [inline libevmcore/ExpressionClasses.h]: negariot -> negation ?

**Pull request merged**